### PR TITLE
Parse attributes on local function declarations

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #38801

Attributes on local function parameters or type parameters are already parsed as expected but they simply give errors. This PR adds an AttributeLists property to LocalFunctionStatementSyntax and parses the attributes into that property.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> This PR adds an AttributeLists property to LocalFunctionDeclarationSyntax and parses the attributes into that property.

Definitely a way to go!   Alternatively, previously, i made it so that *any* MemberDeclarationSyntax could have Attributes.  If the impl is *not* too complex here, you might want to consider making it so that any *statement* could have attributes on it, and that we just bail on them for anything but local functions.

This has hte virtue of embedding them in our syntactic model consistently, and making it much easier to represent in the tree for editing/recovery scenario.

If this is complex, do not do it.  BUt def something to maybe consider :)

File: src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4 [hunk 1]
```diff
@@ -472,7 +472,7 @@ local_declaration_statement
   ;
 
 local_function_statement
-  : modifier* type identifier_token type_parameter_list? parameter_list type_parameter_constraint_clause* (block | (arrow_expression_clause ';'))
+  : attribute_list* modifier* type identifier_token type_parameter_list? parameter_list type_parameter_constraint_clause* (block | (arrow_expression_clause ';'))
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4]: this makes me happy :D

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -6600,7 +6605,7 @@ private bool IsPossibleLocalDeclarationStatement(bool allowAnyExpression)
             }
 
             tk = this.CurrentToken.ContextualKind;
-            if (IsAdditionalLocalFunctionModifier(tk) &&
+            if (IsAdditionalLocalFunctionModifier(tk) || tk == SyntaxKind.OpenBracketToken &&
                 (tk != SyntaxKind.AsyncKeyword || ShouldAsyncBeTreatedAsModifier(parsingStatementNotDeclaration: true)))
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: i find this really hard to parse.  consider more parens... or extracting into helper function (where these are doc'ed statements), or making locals with comments on them.  

--

specifically, i could not answer why you made this change (i.e. why the openbracket change).  On top of not understanding that, the code also feels actively confusing, making it more difficult to understand why you'd make that change.  Once code gets to that point, simplification and documentation are super helpful.  Thanks!

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: (oh, i'm also an idiot since i read this as `Open Brace` not `Open Bracket`.  But the above points still stand).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -7177,8 +7182,9 @@ private bool IsPossibleStatement(bool acceptAccessibilityMods)
                 case SyntaxKind.InternalKeyword:
                 case SyntaxKind.ProtectedKeyword:
                 case SyntaxKind.PrivateKeyword:
+                // could be a local function with attributes
+                case SyntaxKind.OpenBracketToken:  // PROTOTYPE reuse this or always allow the OpenBracketToken?
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: i would always allow it.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -8134,6 +8140,9 @@ private StatementSyntax ParseLocalDeclarationStatement(SyntaxToken awaitKeywordO
             }
             bool canParseAsLocalFunction = usingKeyword == null;
 
+            var attributes = _pool.Allocate<AttributeListSyntax>();
+            ParseAttributeDeclarations(attributes);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: should have a feature-flag/version-check right?

File: src/Compilers/CSharp/Portable/Syntax/LocalFunctionStatementSyntax.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    public partial class SyntaxFactory
+    {
+        public static LocalFunctionStatementSyntax LocalFunctionStatement(
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/LocalFunctionStatementSyntax.cs]: consider doc'ing this is for binary compat.

File: src/Compilers/CSharp/Portable/Syntax/LocalFunctionStatementSyntax.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    }
+}
+
+namespace Microsoft.CodeAnalysis.CSharp.Syntax
+{
+    public partial class LocalFunctionStatementSyntax
+    {
+        public LocalFunctionStatementSyntax Update(SyntaxTokenList modifiers, TypeSyntax returnType, SyntaxToken identifier, TypeParameterListSyntax typeParameterList, ParameterListSyntax parameterList, SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses, BlockSyntax body, ArrowExpressionClauseSyntax expressionBody, SyntaxToken semicolonToken)
+        {
+            if (modifiers != this.Modifiers || returnType != this.ReturnType || identifier != this.Identifier || typeParameterList != this.TypeParameterList || parameterList != this.ParameterList || constraintClauses != this.ConstraintClauses || body != this.Body || expressionBody != this.ExpressionBody || semicolonToken != this.SemicolonToken)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/LocalFunctionStatementSyntax.cs]: could you not sit this on the generated `Update` mthod, just passing in `this.AttributeLists` to it?

**Comment by CyrusNajmabadi** [comment]
Looking good!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I think extracting part of the condition into a named local (something like `isPossibleLocalFunction`?) could be helpful.

File: src/Compilers/CSharp/Portable/Syntax/LocalFunctionStatementSyntax.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Syntax/LocalFunctionStatementSyntax.cs]: Ah yeah, it's probably better to copy paste the signature and modify it, not necessarily bring the whole implementation along with it :smile:

File: src/Compilers/CSharp/Portable/Syntax/LocalFunctionStatementSyntax.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Syntax/LocalFunctionStatementSyntax.cs]: Do you mean a `//` doc for the maintainers of this file?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Yes, thanks. Will make sure to `// PROTOTYPE` this.

**Comment by RikkiGibson** [comment]
Thanks for your feedback, will look into adding AttributeLists to StatementSyntax. (instinctively I wonder if that would be a significant amount of churn in LanguageParser?) Would this make it so if we have some error recovery scenario where maybe some local declaration doesn't parse as a local function, but it happens to have attributes, we could give better diagnostics?

**Comment by jcouv** [comment]
I don't think this PR should target `master`. Will probably need a feature branch.

**Comment by RikkiGibson** [comment]
I used the wrong naming convention for the base branch 🤦‍♂

**Comment by RikkiGibson** [comment]
Apparently I have broken some VB scripting tests somehow.

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Test/Syntax/Parsing/LocalFunctionParsingTests.cs [hunk 7]
```diff
@@ -302,6 +302,119 @@ void M6()
             EOF();
         }
 
+        [Fact]
+        [WorkItem(38801, "https://github.com/dotnet/roslyn/issues/38801")]
+        public void LocalFunctionAttribute()
+        {
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LocalFunctionParsingTests.cs]: >{ [](start = 8, length = 1)

Please test:
```C#
void M()
{
    [A] object Local() => 0;
}
```
```C#
void M()
{
    [A][B] void Local() { }
}
```
```C#
void M()
{
    [A]
}
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I went ahead and added the feature flag check for this scenario.

File: src/Compilers/CSharp/Test/Syntax/Parsing/LocalFunctionParsingTests.cs [hunk 8]
```diff
@@ -302,6 +302,376 @@ void M6()
...
+
+            tree.GetDiagnostics().Verify(
+                // (5,6): error CS1513: } expected
+                //     {
+                Diagnostic(ErrorCode.ERR_RbraceExpected, "").WithLocation(5, 6),
+                // (10,6): error CS1513: } expected
+                //     {
+                Diagnostic(ErrorCode.ERR_RbraceExpected, "").WithLocation(10, 6),
+                // (11,11): error CS0116: A namespace cannot directly contain members such as fields or methods
+                //         [A]
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LocalFunctionParsingTests.cs]: Please split the test into three separate tests because it looks like we're not treating this attribute as appearing within a method.

**Comment by RikkiGibson** [comment]
Could I please get a second review @dotnet/roslyn-compiler

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 4 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Did you add tests for the language feature flag?

---
In reply to: [328751045](https://github.com/dotnet/roslyn/pull/38808#discussion_r328751045) [](ancestors = 328751045)

**Comment by jaredpar** [comment]
>         private VariableDeclaratorSyntax ParseVariableDeclarator(

What happens if an attribute is applied to a standard local variable? Looks like the parser will parse out the attribute but then not attach it to any other node. Won't that cause a consistency issue in the parse tree? Feel like we should have tests for this. 

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4145 in 2c251ee. [](commit_id = 2c251eeb198c7877a601fa1bec39cc1c5c8b7d58, deletion_comment = False)

**Comment by jaredpar** [comment]
Review pass complete (iteration 9)

---
In reply to: [535713832](https://github.com/dotnet/roslyn/pull/38808#issuecomment-535713832) [](ancestors = 535713832)

**Comment by RikkiGibson** [comment]
We have tests which include an unexpected attribute on a local variable declaration, but I can go into more detail in them to show exactly how to access the unexpected attribute if you’d like.

**Comment by jaredpar** [comment]
Where are the tests? If the tests do exist why are they not tripping up the new error you just added about requiring a new version of C#? I would presume the old tests are still using langversion 8.0 hence they should trip up the new error. 

Also is there a test where an attribute is used on a local declaration statement that has multiple locals?


---
In reply to: [536000459](https://github.com/dotnet/roslyn/pull/38808#issuecomment-536000459) [](ancestors = 536000459)

**Comment by RikkiGibson** [comment]
I have a few local declaration with attribute parsing tests in here but some already exist such as `ParserErrorMessageTests.AttributeInMethodBody`.

The LangVersion error is only going to happen if the attribute was on a local function. The diagnostics for putting an attribute on any statement besides LocalFunctionStatement aren't expected to change.

File: src/Compilers/CSharp/Test/Syntax/Parsing/LocalFunctionParsingTests.cs [hunk 9]
```diff
@@ -648,6 +648,91 @@ void M()
...
+        [Fact]
+        [WorkItem(38801, "https://github.com/dotnet/roslyn/issues/38801")]
+        public void LocalFunctionAttribute_Error_LocalVariable_MultipleDeclarators()
+        {
+            var tree = UsingTree(@"
+class C
+{
+    void M()
+    {
+        [A] object local1, local2;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LocalFunctionParsingTests.cs]: What hapens if the attribute is before the second local?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Syntax/Parsing/LocalFunctionParsingTests.cs]: When the attribute comes after the type the parser believes it's some kind of bad array declarator.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Will revisit this when it comes time to remove the prototype comments before merging to master.

**Comment by RikkiGibson** [comment]
I ended up sticking with AttributeLists on the LocalFunctionStatement and the main reason is that Modifiers is also declared on LocalFunctionStatement (not inherited) and it feels like the two should be consistent with each other.

**Comment by CyrusNajmabadi** [comment]
> I ended up sticking with AttributeLists on the LocalFunctionStatement and the main reason is that Modifiers is also declared on LocalFunctionStatement (not inherited) and it feels like the two should be consistent with each other.

I agree.  Though note this is something that can change in the future.  Fields can "move up" the inheritance hierachy without problems.  As i mentioned before, i did this for modifiers/attributes on MemberDeclarationSyntax.

If you're ok with it, i may try to make the attributes-approach more generalized here in your branch.

**Comment by RikkiGibson** [comment]
That makes sense, if you send the PR I am happy to review :smile:
